### PR TITLE
blockchain import_route

### PR DESCRIPTION
**Pull request by debris**
cc @tomusdrw . You probably need it :)

### CODE REVIEW

File: ethcore/src/blockchain/import_route.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+use util::hash::H256;
+use blockchain::block_info::{BlockInfo, BlockLocation};
+
+/// Import route for newly inserted block.
+#[derive(Debug, PartialEq)]
+pub struct ImportRoute {
+	/// Blocks that were invalidated by new block.
+	pub invalidated_blocks: Vec<H256>,
+	/// Blocks that were validted by new block.
```
Comment (tomusdrw) [inline ethcore/src/blockchain/import_route.rs]: typo

**Comment by tomusdrw** [comment]
In `BlockLocation::BranchBecomingCanonChain` you have `route` and `old_route` and then it becomes `validated_blocks` and `invalidated_blocks` contained in `ImportRoute`. I'm not convinced about those names.

What about something like `InsertResult` (method is called `insert_block`) with `canon_blocks` and `retracted_blocks` or even reusing same enum instead of `BlockLocation`.
```rust
enum InsertResult {
  CanonChain(route),
  Branch(route),
  BranchBecameCanonChain(retracted_route/invalidated_route, route)
}
```

**Comment by debris** [comment]
`canon_blocks` / `retracted_blocks` sounds good for me.

But regarding this enum:

```rust
enum InsertResult/InsertLocation {
  CanonChain(route),
  Branch(route),
  BranchBecameCanonChain(retracted_route/invalidated_route, route)
}
```

there is no reason to have route for branch / canonchain. Branch does not have a route at all, and for canon chain, route == inserted block hash. I would prefer to keep `InsertResult/InsertLocation` structures separate, cause it's easier to use them.

**Comment by tomusdrw** [comment]
Agreed. What about `ImportRoute` name?

**Comment by debris** [comment]
I'm open for other ideas. Current name comes from cpp implementation.

**Comment by tomusdrw** [comment]
I vote for keeping it consistent with cpp or calling it `InsertResult`. Up to you.

File: ethcore/src/blockchain/block_info.rs [hunk 2]
```diff
@@ -42,6 +44,8 @@ pub enum BlockLocation {
 		/// Hash of the newest common ancestor with old canon chain.
 		ancestor: H256,
 		/// Hashes of the blocks between ancestor and this block.
-		route: Vec<H256>
+		route: Vec<H256>,
```
Comment (gavofyork) [inline ethcore/src/blockchain/block_info.rs]: `new_canon`

File: ethcore/src/blockchain/block_info.rs [hunk 3]
```diff
@@ -42,6 +44,8 @@ pub enum BlockLocation {
...
+		/// Hashes of the blocks which were invalidated.
+		old_route: Vec<H256>,
```
Comment (gavofyork) [inline ethcore/src/blockchain/block_info.rs]: `retracted`

**Comment by gavofyork** [comment]
prefer snappier naming than the asymmetric `route` and `old_route`.

**Pull request merged**